apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs

```diff
@@ -530,297 +533,605 @@ public Flow(long maxByteCount, long maxElementCount, Action<Task> registerTaskFn
+                Task initTask = _pull.WriteAsync(new StreamingPullRequest
+                {
+                    SubscriptionAsSubscriptionName = _subscriptionName,
+                    StreamAckDeadlineSeconds = _modifyDeadlineSeconds
+                });
+                Add(initTask, Next(true, () => HandlePullMoveNext(initTask)));
+            }
+
+            private bool HandleRpcFailure(Exception e)
+            {
+                if (e != null)
+                {
+                    if (e.As<RpcException>()?.IsRecoverable() ?? false)
+                    {
+                        // Recoverable RPC error, stop and restart pull.
+                        StopStreamingPull();
+                        // Increase backoff internal and start stream again.
+                        // If stream-pull fails repeatly, increase the delay, up to a maximum of 30 seconds.
+                        _pullBackoff = _pullBackoff is TimeSpan backoff ? TimeSpan.FromTicks(backoff.Ticks * 2) : TimeSpan.FromSeconds(0.5);
+                        if (_pullBackoff.Value > TimeSpan.FromSeconds(30))
                         {
-                            streamingPullCts.Cancel();
+                            _pullBackoff = TimeSpan.FromSeconds(30);
                         }
-                        _qEvent.Set();
-                        // Await until pull and push have both stopped.
-                        // If pull of push is faulted/cancelled, this will throw.
-                        await _taskHelper.ConfigureAwait(_taskHelper.WhenAll(pushTask, pullTask));
+                        StartStreamingPull();
+                        return true;
                     }
-                    catch (Exception e0) when (
-                        e0.AllExceptions().All(e => e.IsCancellation() || (e.As<RpcException>()?.IsRecoverable() ?? false)))
+                    else
                     {
-                        // If all errors are recoverable, then do nothing here; restart pull on loop.
+                        // Unrecoverable error; throw it.
+                        throw e.FlattenIfPossible();
                     }
-                    finally
+                }
+                return false;
+            }
+
+            // Pull-stream is ready; call MoveNext to wait for messages.
+            private void HandlePullMoveNext(Task initTask)
+            {
+                // Check if the init write failed.
+                if (initTask != null && HandleRpcFailure(initTask.Exception))
+                {
+                    return;
+                }
+                // Check if pulls need throttling due to push queues being too full, or too slow to push.
+                bool throttle = _extendQueue.TryPeek(out var qItem) && _extendThrottleLow >= qItem.Time;
+                if (!throttle)
+                {
+                    int totalQueueCount = _pushInFlight + _extendQueue.Count;
```

Comment:
Why do these not need to be accessed in the lock?

Reply:
Because they are not used concurrently at all. They are only used within the continuation-based state-machine.
_ack/_nack queues are used concurrently, so are locked.

Reply:
Maybe it's worth adding comments for the fields, to make correct usage clear?

Reply:
There is a comment on _lock stating what needs locking. Is that sufficient?

Reply:
Sure :)